backend/mailru/mailru.go

```diff
@@ -0,0 +1,2356 @@
+		switch strings.ToLower(strings.TrimSpace(flag)) {
+		case "gzip":
+			// This backend mimics the official client which never sends the
+			// "Accept-Encoding: gzip" header. However, enabling compression
+			// might be good for performance.
+			// Use this quirk to investigate the performance impact.
+			// Remove this quirk if perfomance does not improve.
+			q.gzip = true
+		case "insecure":
+			// The mailru disk-o protocol is not documented. To compare HTTP
+			// stream against the official client one can use Telerik Fiddler,
+			// which introduces a self-signed certificate. This quirk forces
+			// the Go http layer to accept it.
+			// Remove this quirk when the backend reaches maturity.
+			q.insecure = true
+		case "binlist":
+			// The official client sometimes uses a so called "bin" protocol,
+			// implemented in the listBin file system method below. This method
+			// is generally faster than non-recursive listM1 but results in
+			// sporadic deserialization failures if total size of tree data
+			// approaches 8Kb (?). The recursive method is normally disabled.
+			// This quirk can be used to enable it for further investigation.
+			// Remove this quirk when the "bin" protocol support is complete.
+			q.binlist = true
+		case "atomicmkdir":
+			// At the moment rclone requires Mkdir to return success if the
+			// directory already exists. However, such programs as borgbackup
+			// or restic use mkdir as a locking primitive and depend on its
+			// atomicity. This quirk is a workaround. It can be removed
+			// when the above issue is investigated.
+			q.atomicmkdir = true
+		case "retry400":
+			// This quirk will help in troubleshooting a very rare "Error 400"
+			// issue. It can be removed if the problem does not show up
+			// for a year or so. See the below issue:
+			// https://github.com/ivandeex/rclone/issues/14
+			q.retry400 = true
+		default:
+			// Just ignore all unknown flags
+		}
+	}
+}
+
+// Note: authorize() is not safe for concurrent access as it updates token source
+func (f *Fs) authorize(ctx context.Context, force bool) (err error) {
+	var t *oauth2.Token
+	if !force {
+		t, err = oauthutil.GetToken(f.name, f.m)
+	}
+
+	if err != nil || !tokenIsValid(t) {
+		fs.Infof(f, "Valid token not found, authorizing.")
+		ctx := oauthutil.Context(f.cli)
+		t, err = oauthConfig.PasswordCredentialsToken(ctx, f.opt.Username, f.opt.Password)
+	}
+	if err == nil && !tokenIsValid(t) {
+		err = errors.New("Invalid token")
+	}
+	if err != nil {
+		return errors.Wrap(err, "Failed to authorize")
+	}
+
+	if err = oauthutil.PutToken(f.name, f.m, t, false); err != nil {
+		return err
+	}
+
+	// Mailru API server expects access token not in the request header but
+	// in the URL query string, so we must use a bare token source rather than
+	// client provided by oauthutil.
+	//
+	// WARNING: direct use of the returned token source triggers a bug in the
+	// `(*token != *ts.token)` comparison in oauthutil.TokenSource.Token()
```

Comment:
Do you want to fix up oauthutil?  You could do that in another commit in this PR if you wanted? Or make a new issue about it?

Reply:
Will submit an issue.